include/aikido/control/detail/KinematicSimulationJointCommandExecutor-impl.hpp

```diff
@@ -0,0 +1,209 @@
+#include <dart/common/Console.hpp>
+
+#include "aikido/common/memory.hpp"
+
+namespace aikido {
+namespace control {
+
+//==============================================================================
+extern template class KinematicSimulationJointCommandExecutor<
+    ExecutorType::POSITION>;
+
+extern template class KinematicSimulationJointCommandExecutor<
+    ExecutorType::VELOCITY>;
+
+//==============================================================================
+template <ExecutorType T>
+KinematicSimulationJointCommandExecutor<T>::
+    KinematicSimulationJointCommandExecutor(
+        ::dart::dynamics::MetaSkeletonPtr metaskeleton)
+  : JointCommandExecutor<T>(
+      checkNull(metaskeleton)->getDofs(),
+      std::set<ExecutorType>{ExecutorType::STATE})
+  , mInProgress{false}
+  , mPromise{nullptr}
+  , mMutex{}
+{
+}
+
+//==============================================================================
+template <ExecutorType T>
+KinematicSimulationJointCommandExecutor<
+    T>::~KinematicSimulationJointCommandExecutor()
+{
+  {
+    std::lock_guard<std::mutex> lock(mMutex);
+    DART_UNUSED(lock); // Suppress unused variable warning
+
+    if (mInProgress)
+    {
+      mInProgress = false;
+      mPromise->set_exception(
+          std::make_exception_ptr(std::runtime_error("Command canceled.")));
+    }
+    this->stop();
+  }
+}
+
+//==============================================================================
+template <ExecutorType T>
+std::future<int> KinematicSimulationJointCommandExecutor<T>::execute(
+    const std::vector<double>& command,
+    const std::chrono::duration<double>& timeout)
+{
+  auto promise = std::promise<int>();
+
+  if (command.size() != this->getDofs().size())
+  {
+    promise.set_exception(std::make_exception_ptr(
+        std::runtime_error("DOF of command does not match DOF of joints.")));
+    return promise.get_future();
+  }
+
+  std::lock_guard<std::mutex> lock(mMutex);
+  DART_UNUSED(lock); // Suppress unused variable warning
+  if (mInProgress)
+  {
+    // Overwrite previous command
+    mCommand.clear();
+    mPromise->set_exception(
+        std::make_exception_ptr(std::runtime_error("Command canceled.")));
+  }
+
+  mPromise.reset(new std::promise<int>());
+
+  mCommand = command;
+  mInProgress = true;
+  mExecutionStartTime = std::chrono::system_clock::now();
+  mTimeout = timeout;
+
+  // Set start positions
+  mStartPosition.clear();
+  for (std::size_t i = 0; i < this->getDofs().size(); ++i)
+  {
+    auto dof = this->getDofs()[i];
+    mStartPosition.push_back(dof->getPosition());
+  }
```

Comment:
Nit: it seems like we're calling `getDofs()` twich for each loop iteration, which could maybe be fixed by referencing `mDofs` directly or calling it once before the loop and referencing the local variable.

Reply:
Same issue below.